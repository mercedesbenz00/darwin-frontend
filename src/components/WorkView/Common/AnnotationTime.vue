<template>
  <div
    v-tooltip="'Total time the annotator has spent on this image'"
    class="annotation-time"
  >
    <img
      class="annotation-time__icon"
      src="/static/imgs/clock.svg"
    >
    <div class="annotation-time__clock">
      {{ time }}
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'

import { formatDurationAsTimer } from '@/utils'

/**
 * Shows total annotation time on a task, as a HH:MM:SS time display with an icon in front
 */
@Component({ name: 'annotation-time' })
export default class AnnotationTime extends Vue {
  @Prop({ required: false, default: null, type: Number })
  timeInSeconds!: number | null

  get time (): string {
    const { timeInSeconds } = this
    return formatDurationAsTimer(timeInSeconds || 0)
  }
}
</script>

<!-- eslint-disable vue-scoped-css/enforce-style-type -->
<style lang="scss">
.annotation-time {
  @include row--center;

}

.annotation-time__icon {
  margin-right: 5px;
  padding-bottom: 3px;
}

.annotation-time__clock {
  @include typography(md-1, default, 500);
}
</style>
