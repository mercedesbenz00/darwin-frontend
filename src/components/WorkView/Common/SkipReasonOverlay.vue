<template>
  <div class="overlay">
    {{ text }}
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'

/**
 * Used to style a div overlayed on top of workview center,
 * which renders the skipped/discarded/rejected/archived message.
 *
 * To maintain reusability between new and classic workflow, this accepts a text prop.
 *
 * TODO:
 *
 * Once classic worview is done, we should use selected stage instance
 * and dataset item from state, to infer what should be rendered.
 *
 * We should also pass styling to workflow.
 */
@Component({ name: 'skip-reason-overlay' })
export default class SkipReasonOverlay extends Vue {
  @Prop({ required: true, type: String })
  text!: string
}

</script>

<style lang="scss" scoped>
.overlay {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  padding: 10px 20px;
  border-radius: 5px;
  border: 2px solid $colorWhite;
  @include typography(xl, default, bold);
  background: rgba(255, 40, 117, 0.5);
  color: $colorWhite;
}
</style>
