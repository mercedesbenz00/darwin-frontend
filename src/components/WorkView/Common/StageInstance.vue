<template>
  <status-button
    class="instance"
    :class="{'instance--selected': isSelected}"
    :type="instance.type"
    inverted
    @click="$emit('click')"
  >
    <team-member-avatar
      v-if="assignee"
      class="instance__avatar"
      :member="assignee"
    />
  </status-button>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import { Getter, State } from 'vuex-class'

import TeamMemberAvatar from '@/components/Common/Avatar/V1/ResponsiveTeamMemberAvatar.vue'
import StatusButton from '@/components/WorkView/Common/StatusButton/V2/StatusButton.vue'
import { MembershipPayload, WorkflowStagePayload } from '@/store/types'

@Component({ name: 'stage-instance', components: { StatusButton, TeamMemberAvatar } })
export default class StageInstance extends Vue {
  @Prop({ required: true, type: Object as () => WorkflowStagePayload })
  instance!: WorkflowStagePayload

  @State(state => state.workview.selectedStageInstance)
  selectedInstance!: WorkflowStagePayload | null

  get isSelected (): boolean {
    const { instance, selectedInstance } = this
    return !!selectedInstance && selectedInstance.id === instance.id
  }

  @Getter('relevantTeamMemberships', { namespace: 'team' })
  memberships!: MembershipPayload[]

  get assignee (): MembershipPayload | null {
    const { instance, memberships } = this
    return memberships.find(m => m.user_id === instance.assignee_id) || null
  }
}
</script>
<style lang="scss" scoped>
.instance {
  border-width: 2px;
}

.instance.status-button,
.instance:hover.status-button {
  @include workflow-status-background-color;
  color: $colorWhite;
}

</style>
