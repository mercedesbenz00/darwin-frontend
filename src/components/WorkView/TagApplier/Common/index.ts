export { default as TagApplierHeader } from './TagApplierHeader/TagApplierHeader.vue'
export { default as TagApplierInput } from './TagApplierInput/TagApplierInput.vue'
export { default as TagApplierList } from './TagApplierList/TagApplierList.vue'
export { default as TagApplierCreate } from './TagApplierCreate/TagApplierCreate.vue'
