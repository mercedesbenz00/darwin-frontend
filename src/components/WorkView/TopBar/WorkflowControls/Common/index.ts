export { default as ActionHistory } from './ActionHistory/ActionHistory.vue'
export { default as ActionHistoryTrigger } from './ActionHistory/ActionHistoryTrigger.vue'
export { default as V2ActionHistoryTrigger } from './ActionHistory/V2/ActionHistoryTrigger.vue'
export { default as ContinueButton } from './ContinueButton/ContinueButton.vue'
export { default as DiscardButton } from './DiscardButton/DiscardButton.vue'
export { default as StageArchiveButton } from './StageArchiveButton/StageArchiveButton.vue'
export { default as StageRejectButton } from './StageRejectButton/StageRejectButton.vue'
export { default as StageSkipButton } from './StageSkipButton/V1/StageSkipButton.vue'
export { default as V2StageSkipButton } from './StageSkipButton/V2/StageSkipButton.vue'
