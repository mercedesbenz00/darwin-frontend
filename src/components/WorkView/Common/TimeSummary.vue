<template>
  <div class="summary">
    <template v-if="summary && currentDurationData.length > 0">
      <div
        v-for="data in currentDurationData"
        :key="data.key"
        class="summary-item"
      >
        <status-button
          class="summary-item__avatar"
          :type="data.type"
        >
          <team-member-avatar
            v-if="data.membership"
            :member="data.membership"
          />
        </status-button>
        <div class="summary-item__name">
          {{ data.name }}
        </div>
        <annotation-time
          class="summary-item__time"
          :time-in-seconds="data.duration"
        />
      </div>
    </template>
    <template v-else>
      No time tracked yet.
    </template>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import { State } from 'vuex-class'

import TeamMemberAvatar from '@/components/Common/Avatar/V1/ResponsiveTeamMemberAvatar.vue'
import AnnotationTime from '@/components/WorkView/Common/AnnotationTime.vue'
import {
  DatasetItemTimeSummaryPayload,
  MembershipPayload,
  WorkflowStagePayload,
  StageType
} from '@/store/types'
import { getFullName } from '@/utils'

import StatusButton from './StatusButton/V1/StatusButton.vue'

type SummaryItem = {
  duration: number
  key: string
  membership?: MembershipPayload
  name: string
  type: WorkflowStagePayload['type']
}

const buildSummaryItem = (
  stage: WorkflowStagePayload,
  memberships: MembershipPayload[],
  userId: number,
  duration: number,
  type: Exclude<
    StageType,
    StageType.Dataset | StageType.Discard | StageType.ConsensusEntrypoint |
    StageType.ConsensusTest | StageType.Logic
  >
): SummaryItem => {
  const key = `${userId}-${stage.id}`
  const membership = memberships.find(m => m.user_id === userId)
  const name = membership ? getFullName(membership) : 'Unknown user'
  return { duration, type, key, membership, name }
}

/**
 * In charge of rendering of a time summary for the specified dataset item
 *
 * Loading is done separately from this component, in `TimeSummaryLoader.ts`
 */
@Component({
  name: 'time-summary',
  components: { AnnotationTime, StatusButton, TeamMemberAvatar }
})
export default class TimeSummary extends Vue {
  @Prop({ required: true, type: Object as () => WorkflowStagePayload })
  stage!: WorkflowStagePayload

  get summary (): DatasetItemTimeSummaryPayload | null {
    const summaries =
      this.$store.state.workview.datasetItemTimeSummaries as DatasetItemTimeSummaryPayload[]
    return summaries.find(s => s.dataset_item_id === this.stage.dataset_item_id) || null
  }

  @State(state => state.team.memberships)
  memberships!: MembershipPayload[]

  /**
   * Returns formatted duration data for this stage, structured so it's easier to render.
   *
   * If the stage is assigned, this array needs to always return at least one
   * item, which is the duration tracked for the stage assignee (0 default).
   */
  get currentDurationData (): SummaryItem[] {
    const { memberships, stage, summary } = this

    const data: SummaryItem[] = []

    const sourceData = summary ? summary.current_workflow.per_stage_per_user : []

    sourceData.filter(d => d.stage_id === stage.id).forEach(d => {
      if (d.type === 'complete') { return }
      if (!d.user_id) { return }
      if (d.duration === 0) { return }
      const item = buildSummaryItem(stage, memberships, d.user_id, d.duration, d.type)
      data.push(item)
    })

    const assigneeHasData =
      data.some(d => d.membership && d.membership.user_id === stage.assignee_id)

    if (stage.assignee_id && !assigneeHasData) {
      const item = buildSummaryItem(stage, memberships, stage.assignee_id, 0, stage.type)
      data.push(item)
    }

    return data
  }
}
</script>

<style lang="scss" scoped>
.summary {
  background: $colorWhite;
  border-radius: 5px;
  color: $color90Black;
  padding: 18px;
  box-shadow: $dropShadowNeutral;
  width: 200px;
  @include typography(md-1, default, 500);
}

.summary-item {
  display: grid;
  grid-auto-columns: max-content;
  grid-gap: 10px;
  grid-template-areas: "avatar name" "duration duration";
  align-content: center;
  margin-bottom: 20px;
}

.summary-item__avatar {
  height: 30px;
  width: 30px;
  grid-area: avatar;
}

.summary-item__name {
  grid-area: name;
  align-self: center;
}

.summary-item__time {
  grid-area: duration;
  justify-content: left;
  justify-self: left;
  background: $colorAliceBlue;
  padding: 6px;
  border-radius: 3px;
}

.summary-item__time :deep(.annotation-time__icon) {
  width: 25px;
  height: 28px;
}
</style>
