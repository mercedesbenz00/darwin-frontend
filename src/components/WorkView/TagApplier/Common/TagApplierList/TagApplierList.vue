<template>
  <div class="tag-applier__list">
    <recycle-scroller
      ref="recycleScroller"
      v-if="itemsLength"
      class="tag-applier__list__content"
      :buffer="buffer"
      :item-size="itemHeight"
      :items="parsedItems"
      direction="vertical"
      key-field="label"
      emit-update
      @resize="onResize"
      @update="onUpdate"
    >
      <template #default="{ item, index }">
        <list-element-v2
          v-tooltip="{
            content: `Add '${item.label}' tag annotation`
          }"
          tabindex="1"
          :selected="focused && selectedIdx === Number(index)"
          @click="$emit('create', item.label)"
        >
          <badge
            :label="item.label"
            :color="item.color"
            size="medium"
            :high-contrast="focused && selectedIdx === Number(index)"
            no-capitalize
            no-tooltip
          />
          <keycap
            v-if="focused && selectedIdx === Number(index)"
            size="small"
          >
            â†µ
          </keycap>
        </list-element-v2>
      </template>
    </recycle-scroller>
    <span
      v-else
      class="tag-applier__list__no-content"
    >
      <span
        v-if="duplicate"
        class="tag-applier__list__no-content__new-badge"
      >
        Tag {{ keyword }} already exists
      </span>
      <span
        v-else-if="keyword"
        class="tag-applier__list__no-content__new-badge"
      >
        Tag {{ keyword }} not available
      </span>
      <span v-else>No tags available</span>
    </span>
  </div>
</template>

<script lang="ts">
import { sortBy } from 'lodash'
import {
  defineComponent,
  computed,
  PropType,
  ref,
  Ref,
  SetupContext,
  watch
} from 'vue'
import { RecycleScroller } from 'vue-virtual-scroller'

import { Badge, BadgeType } from '@/components/Common/Badge'
import Keycap from '@/components/Common/Keycap/Keycap.vue'
import ListElementV2 from '@/components/Common/ListElements/ListElementV2/ListElementV2.vue'
import { useFocus, useHotkeysNavigation, shouldScroll, useTheme } from '@/composables'
import { getIdNext } from '@/utils'

export default defineComponent({
  name: 'TagApplierList',
  components: { Badge, ListElementV2, Keycap, RecycleScroller },
  props: {
    keyword: { type: String, required: true },
    items: { type: Array as PropType<BadgeType[]>, required: true },
    direction: { type: String as PropType<'asc' | 'desc'>, default: 'asc' },
    duplicate: { type: Boolean, default: false }
  },
  setup (props, { emit }: SetupContext) {
    const theme = useTheme()
    const recycleScroller = ref<any>()
    const focused: Ref<boolean> = ref(false)
    const selectedIdx = ref(0)
    const bufferThreshold: Ref<number> = ref(2)
    const recycleScrollerHeight: Ref<number> = ref(1)
    const firstVisibleIdx: Ref<number> = ref(0)
    const lastVisibleIdx: Ref<number> = ref(0)

    const itemHeight = computed((): number => {
      return 28 * theme.getCurrentScale()
    })

    const visibleItems = computed((): number => {
      return Math.round(recycleScrollerHeight.value / itemHeight.value)
    })

    const itemsLength = computed((): number => {
      return props.items ? props.items.length : 0
    })

    const buffer = computed((): number => {
      return Math.ceil(itemHeight.value * bufferThreshold.value)
    })

    /**
     * Check if keyword already exists as an annotation class
     */
    const parsedItems = computed((): BadgeType[] => {
      const sorted = sortBy(props.items, ({ label }) => label)
      return props.direction === 'asc' ? sorted : sorted.reverse()
    })

    /**
     * When the parsedItem get updated (eg; due to filtering)
     * reset the selectedIdx to avoid accessing a index of parsedItems
     * exceeding its length
     */
    watch(() => parsedItems.value, () => {
      setTimeout(() => (selectedIdx.value = 0), 0)
    })

    const scrollToItem = (index: number = 0): void => {
      if (recycleScroller.value && 'scrollToItem' in recycleScroller.value) {
        recycleScroller.value.scrollToItem(index)
      }
    }

    useFocus('tag-applier__list', (value: boolean) => {
      focused.value = value
    })

    useHotkeysNavigation({
      condition: focused,
      handlers: {
        increase: () => {
          const newIdx = getIdNext(selectedIdx.value, itemsLength.value, true)
          selectedIdx.value = newIdx
          const scrollToIdx = shouldScroll({
            idx: newIdx,
            next: true,
            length: itemsLength.value,
            firstIdx: firstVisibleIdx.value,
            lastIdx: lastVisibleIdx.value,
            visibleItems: visibleItems.value
          })
          if (scrollToIdx >= 0) { scrollToItem(scrollToIdx) }
        },
        decrease: () => {
          const newIdx = getIdNext(selectedIdx.value, itemsLength.value, false)
          selectedIdx.value = newIdx
          const scrollToIdx = shouldScroll({
            idx: newIdx,
            next: false,
            length: itemsLength.value,
            firstIdx: firstVisibleIdx.value,
            lastIdx: lastVisibleIdx.value,
            visibleItems: visibleItems.value
          })
          if (scrollToIdx >= 0) { scrollToItem(scrollToIdx) }
        },
        submit: () => {
          emit('create', parsedItems.value[selectedIdx.value]?.label)
          emit('close') // used in case tag-applier mini props it's on
        },
        close: () => {
          emit('close') // used in case tag-applier mini props it's on
        }
      },
      name: 'tag applier list',
      special: true
    })

    const onResize = (): void => {
      if (recycleScroller.value && recycleScroller.value.$el &&
        'offsetHeight' in recycleScroller.value.$el) {
        recycleScrollerHeight.value = recycleScroller.value.$el.offsetHeight
      }
    }

    const onUpdate = (startIdx: number, endIdx: number): void => {
      firstVisibleIdx.value = startIdx === 0
        ? startIdx
        : startIdx + bufferThreshold.value
      lastVisibleIdx.value = endIdx === itemsLength.value
        ? itemsLength.value
        : endIdx - bufferThreshold.value - 1
    }

    return {
      recycleScroller,
      focused,
      selectedIdx,
      recycleScrollerHeight,
      firstVisibleIdx,
      lastVisibleIdx,
      itemHeight,
      visibleItems,
      itemsLength,
      buffer,
      parsedItems,
      shouldScroll,
      scrollToItem,
      onResize,
      onUpdate
    }
  }
})
</script>

<style lang="scss" scoped>
.tag-applier__list {
  position: relative;
  display: flex;
  flex: 1 1 auto;
  height: auto;
  width: 100%;
  padding: 4px;
  border-top: 1px solid $colorBorderLight;
  @include scrollbarV2('vertical');

  &:focus,
  &:active {
    outline: none;
  }

  &__content {
    @include col;
    flex: 1;
    z-index: 2;
    height: 100%;
    max-width: 100%;
    @include scrollbarV2('vertical');
    overflow-x: hidden;

    :deep(.vue-recycle-scroller) {
      .vue-recycle-scroller__item-wrapper {

        .vue-recycle-scroller__item-view {

          &.hover {
            .list-element__content {
              background: $colorNeutralsLight100 !important;
            }
          }

          &.active {
            background: $colorNeutralsLight200;
          }

          .list-element {
            height: 28px;
            padding: 2px;

            &__content {
              .badge {
                @include row--wrap;
                align-items: center;
                width: 100%;

                &__content__label {
                  max-width: 100%;
                  @include ellipsis(1, md);
                }
              }
            }
          }
        }
      }
    }
  }

  &__no-content {
    @include row--center;
    flex: 1;
    min-height: 30px;
    max-width: 100%;
    @include typography(md, inter);
    padding: 16px;
    text-align: center;
    color: $colorContentSecondary;

    &__new-badge {
      max-width: 100%;
      @include ellipsis(2, md);
    }
  }
}
</style>
