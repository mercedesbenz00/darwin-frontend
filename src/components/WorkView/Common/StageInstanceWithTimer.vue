<template>
  <div class="instance-wrapper">
    <stage-instance
      class="instance-wrapper__instance"
      :instance="instance"
    />
    <timer
      v-if="instance.completes_at && nextStatus"
      :completes-at="instance.completes_at"
      :class="`instance-wrapper__timer--${nextStatus}`"
      class="instance-wrapper__timer"
      :show-cancel="false"
    />
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import { State, Getter } from 'vuex-class'

import { Timer } from '@/components/WorkView/Common/Timer'
import {
  WorkflowStagePayload,
  WorkflowTemplatePayload,
  StageActionType,
  DatasetItemStatus,
  StageType
} from '@/store/types'
import { StageActionTypeToStatus } from '@/utils/StageActionTypeToStatus'

import StageInstance from './StageInstance.vue'
import { StageInstanceWithTimerProps } from './types'

@Component({ name: 'stage-instance-with-timer', components: { StageInstance, Timer } })
export default class StageInstanceWithTimer extends Vue implements StageInstanceWithTimerProps {
  @Prop({ required: true, type: Object as () => WorkflowStagePayload })
  instance!: WorkflowStagePayload

  get progressing (): boolean {
    const { instance } = this
    if (instance.metadata.review_status === 'rejected') { return false }
    return true
  }

  @State(state => state.workview.workflowTemplates)
  workflowTemplates!: WorkflowTemplatePayload[]

  get workflowTemplate (): WorkflowTemplatePayload | null {
    const { instance, workflowTemplates } = this
    const template = workflowTemplates.find(
      w => w.workflow_stage_templates.some(t => t.id === instance.workflow_stage_template_id)
    )

    return template || null
  }

  @Getter('stageCurrentAction', { namespace: 'workview' })
  getContinueAction!: (stage: WorkflowStagePayload) => StageActionType

  get nextStatus (): DatasetItemStatus | 'rejected' | StageType | null {
    const { instance } = this
    if (instance.type === StageType.Complete) { return null }
    const nextAction = this.getContinueAction(instance)
    return StageActionTypeToStatus[nextAction]
  }
}
</script>

<style lang="scss" scoped>
.instance-wrapper {
  position: relative;
  height: 100%;
  width: 100%;

  :deep(.instance) {
    height: 100%;
    width: 100%;
  }

  > * {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translateX(-50%) translateY(-50%);
  }

  &__timer {
    @include workflow-status-color;
  }
}
</style>
