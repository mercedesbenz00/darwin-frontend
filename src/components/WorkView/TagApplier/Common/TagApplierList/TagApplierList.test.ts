import { createLocalVue, mount, Wrapper } from '@vue/test-utils'
import VTooltip from 'v-tooltip'
import Vue from 'vue'
import Vuex from 'vuex'

import createTestStore from 'test/unit/createTestStore'

import { BadgeType } from '@/components/Common/Badge'

import TagApplierList from './TagApplierList.vue'

const localVue = createLocalVue()
localVue.use(VTooltip, { defaultHtml: true })
localVue.use(Vuex)

let store: ReturnType<typeof createTestStore>
let wrapper: Wrapper<Vue>

const $theme = {
  getCurrentScale: (): number => 1
}

let propsData: {
  keyword: string,
  items: BadgeType[],
  direction?: 'asc' | 'desc',
  focused?: boolean,
}

beforeEach(() => {
  store = createTestStore()
  propsData = {
    keyword: '',
    items: [
      { label: 'foo' },
      { label: 'bar' },
      { label: 'baz' }
    ]
  }
  const mocks = { $theme }
  wrapper = mount(TagApplierList, { mocks, propsData, localVue, store })
})

afterEach(() => {
  wrapper.destroy()
})

describe('TagApplierList', () => {
  it('should mount', () => {
    expect(wrapper.exists()).toBeTruthy()
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should match snapshot with defined label', async () => {
    await wrapper.setProps({ keyword: 'foo' })
    expect(wrapper).toMatchSnapshot()
  })

  it('should contain list items', () => {
    const listItems = wrapper.findAll('.vue-recycle-scroller__item-view')
    expect(listItems.length).toBe(2) // because of the recycle-scroller
  })
})

describe('when submitting using arrow keys', () => {
  let wrapper: Wrapper<Vue, Element>

  beforeEach(() => {
    propsData = {
      keyword: '',
      items: [
        { label: 'foo' },
        { label: 'bar' },
        { label: 'baz' }
      ]
    }
    const mocks = { $theme }
    wrapper = mount(TagApplierList, { localVue, mocks, propsData, store })
  })

  it('select the next value focused and hitting ARROW DOWN', async () => {
    await wrapper.setData({ focused: true })
    expect((wrapper.vm as any).selectedIdx).toEqual(0)
    await document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowDown' }))
    expect((wrapper.vm as any).selectedIdx).toEqual(1)
  })

  it('do not select the next value not focused and hitting ARROW UP', async () => {
    await wrapper.setData({ focused: false })
    expect((wrapper.vm as any).selectedIdx).toEqual(0)
    document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowDown' }))
    expect((wrapper.vm as any).selectedIdx).toEqual(0)
  })

  it('select the prev value open and hitting ARROW UP', async () => {
    await wrapper.setData({ focused: true })
    expect((wrapper.vm as any).selectedIdx).toEqual(0)
    document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowUp' }))
    expect((wrapper.vm as any).selectedIdx).toEqual(2)
  })

  it('do not select the prev value not focused and hitting ARROW UP', async () => {
    await wrapper.setData({ focused: false })
    expect((wrapper.vm as any).selectedIdx).toEqual(0)
    document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowUp' }))
    expect((wrapper.vm as any).selectedIdx).toEqual(0)
  })

  it('submit the selected value when focused and hitting ENTER', async () => {
    await wrapper.setData({ focused: true })
    document.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter' }))
    expect(wrapper.emitted()?.create).toEqual([['bar']])
  })

  it('not submit the selected value when not focused and hitting ENTER', async () => {
    await wrapper.setData({ focused: false })
    document.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter' }))
    expect(wrapper.emitted()?.create).not.toBeDefined()
  })

  it('submit the selected value when focused and hitting ESCAPE', async () => {
    await wrapper.setData({ focused: true })
    document.dispatchEvent(new KeyboardEvent('keydown', { key: 'Escape' }))
    expect(wrapper.emitted()?.close).toEqual([[]])
  })

  it('not submit the selected value when not focused and hitting ESCAPE', async () => {
    await wrapper.setData({ focused: false })
    document.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter' }))
    expect(wrapper.emitted()?.close).not.toBeDefined()
  })
})
