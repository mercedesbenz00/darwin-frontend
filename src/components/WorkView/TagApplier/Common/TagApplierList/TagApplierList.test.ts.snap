// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagApplierList should match snapshot 1`] = `
<div class="tag-applier__list">
  <div class="vue-recycle-scroller tag-applier__list__content ready direction-vertical">
    <!---->
    <div class="vue-recycle-scroller__item-wrapper" style="min-height: 84px;">
      <div class="vue-recycle-scroller__item-view" style="transform: translateY(0px);"><button tabindex="1" class="list-element has-tooltip" data-original-title="null">
          <div class="list-element__content">
            <!---->
            <div class="badge"><span class="badge__content badge__content--size--medium badge__content--bar" style="background-color: rgba(179, 82, 255, 0.1);"> <span class="badge__content__label badge__content__label--none" style="color: rgb(179, 82, 255);">
      bar
    </span>
              <!----></span>
            </div>
            <!---->
          </div>
          <!---->
        </button></div>
      <div class="vue-recycle-scroller__item-view" style="transform: translateY(28px);"><button tabindex="1" class="list-element has-tooltip" data-original-title="null">
          <div class="list-element__content">
            <!---->
            <div class="badge"><span class="badge__content badge__content--size--medium badge__content--baz" style="background-color: rgba(255, 215, 73, 0.1);"> <span class="badge__content__label badge__content__label--none" style="color: rgb(255, 215, 73);">
      baz
    </span>
              <!----></span>
            </div>
            <!---->
          </div>
          <!---->
        </button></div>
    </div>
    <!---->
    <div data-v-b329ee4c="" tabindex="-1" class="resize-observer"><object aria-hidden="true" tabindex="-1" type="text/html" data="about:blank"></object></div>
  </div>
</div>
`;

exports[`TagApplierList should match snapshot with defined label 1`] = `
<div class="tag-applier__list">
  <div class="vue-recycle-scroller tag-applier__list__content ready direction-vertical">
    <!---->
    <div class="vue-recycle-scroller__item-wrapper" style="min-height: 84px;">
      <div class="vue-recycle-scroller__item-view" style="transform: translateY(0px);"><button tabindex="1" class="list-element has-tooltip" data-original-title="null">
          <div class="list-element__content">
            <!---->
            <div class="badge"><span class="badge__content badge__content--size--medium badge__content--bar" style="background-color: rgba(179, 82, 255, 0.1);"> <span class="badge__content__label badge__content__label--none" style="color: rgb(179, 82, 255);">
      bar
    </span>
              <!----></span>
            </div>
            <!---->
          </div>
          <!---->
        </button></div>
      <div class="vue-recycle-scroller__item-view" style="transform: translateY(28px);"><button tabindex="1" class="list-element has-tooltip" data-original-title="null">
          <div class="list-element__content">
            <!---->
            <div class="badge"><span class="badge__content badge__content--size--medium badge__content--baz" style="background-color: rgba(255, 215, 73, 0.1);"> <span class="badge__content__label badge__content__label--none" style="color: rgb(255, 215, 73);">
      baz
    </span>
              <!----></span>
            </div>
            <!---->
          </div>
          <!---->
        </button></div>
    </div>
    <!---->
    <div data-v-b329ee4c="" tabindex="-1" class="resize-observer"><object aria-hidden="true" tabindex="-1" type="text/html" data="about:blank"></object></div>
  </div>
</div>
`;
